docs(framework) Update how to upgrade to Flower 1.13 documentation

dstripelis opened the pull request:
- Added server and client app examples for versions 1.11+.
- Moved sentence highlighting to latest flower versions.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -135,9 +138,19 @@ to run your project both in the traditional way and in the Flower Next way:
+    app = flwr.client.ClientApp(
+        client_fn=client_fn,
+    )
+
```
```suggestion
```

dstripelis replied:
Resolved.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -160,15 +173,26 @@ to run your project both in the traditional way and in the Flower Next way:
 
 .. code-block:: python
-    :emphasize-lines: 2,9
+    :emphasize-lines: 2,5,11
```
```suggestion
    :emphasize-lines: 2,8
```

dstripelis replied:
Yes, it seems more reasonable to highlight both the `client_fn` and `server_fn` and how each respective `Client` and `Server` object is constructed.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -232,15 +256,30 @@ Simulation in CLI
 
-    # Flower 1.8
+    # Flower v1.11+
+    def client_fn(context: flwr.common.Context):
+        return flwr.client.FlowerClient().to_client()
+
+
+    app = flwr.client.ClientApp(
+        client_fn=client_fn,
+    )
+
+    server_app = flwr.server.ServerApp(
+        config=config,
+        strategy=strategy,
+    )
```
```suggestion
    # [file: client_app.py]
    def client_fn(context: flwr.common.Context):
        return flwr.client.FlowerClient().to_client()


    app = flwr.client.ClientApp(
        client_fn=client_fn,
    )

    # [file: server_app.py]
    def server_fn(context: flwr.common.Context):
        strategy = flwr.server.strategy.FedAvg(...)
        config = flwr.server.ServerConfig(...)
        return flwr.server.ServerAppComponents(strategy=strategy, config=config)
        
    server_app = flwr.server.ServerApp(
        server_fn=server_fn,
    )
```

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -267,35 +306,35 @@ Simulation in CLI
 .. code-block:: bash
 
-    # Flower 1.8
+    # Flower v1.8
```
```suggestion
    # Flower v1.13+
    $ flwr run
    
    # Flower v1.8 - v1.10
```

@jafermarq I'm not familiar with the `flower-simulation` versions. Is this part correct?

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -328,15 +367,15 @@ Simulation in a Notebook
     backend_config = {"client_resources": {"num_cpus": 2, "num_gpus": 0.25}}
 
-    # Flower 1.8
+    # Flower v1.8
```
We need to specify that from `flwr == 1.11` onwards we use the `server_fn` and `client_fn` with context. Refer to the [Use federated learning strategy](file:///Users/csng/code/flower/doc/build/html/tutorial-series-use-a-federated-learning-strategy-pytorch.html) tutorial.

chongshenng reviewed (changes requested):
@dstripelis Thanks for starting this update! There're 2 large parts that need to be updated:
* On the Deployment Engine, we no longer use `flower-server-app` and `flower-client-app`. These need to be removed. Maybe you can work together with @Robert-Steiner since he has PR #4372 which also updates the deployment engine mode. The secure connection part also needs to be updated (e.g. by referring to `doc/source/docker/enable-tls.rst` in #4470).
* On the Simulation Engine, let's update to the latest `server_fn` and `client_fn` usage. We should start either modes by using `flwr run`, and keep `run_simulation` for running in notebooks (less use and emphasized).

dstripelis commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -254,62 +307,120 @@ Simulation in CLI
+        server_fn=server_fn,
+    )
+
+    flwr.simulation.run_simulation(
```
@danielnugraha this is inside the notebook so the simulation has to be started this way; @jafermarq please correct me if there is not consistent.

chongshenng replied:
Yes,  although `flwr run` is the preferred way, we still (have to) use `run_simulation` in notebooks.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -189,6 +219,30 @@ Deployment
+         --insecure \
+         --superlink 127.0.0.1:9092 \
+         --node-config "partition-id=0 num-partitions=2" \
+         --supernode-address 127.0.0.1:9094 \
+         --isolation subprocess
```
```suggestion
         --supernode-address 127.0.0.1:9094
```

`--isolation subprocess` is the default mode, so we can omit this.

dstripelis replied:
Fixed.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -189,6 +219,30 @@ Deployment
+         --insecure \
+         --superlink 127.0.0.1:9092 \
+         --node-config "partition-id=1 num-partitions=2" \
+         --supernode-address 127.0.0.1:9095 \
+         --isolation subprocess
```
```suggestion
         --supernode-address 127.0.0.1:9095
```

`--isolation subprocess` is the default mode, so we can omit this.

dstripelis replied:
Fixed.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -189,6 +219,30 @@ Deployment
+         --supernode-address 127.0.0.1:9095 \
+         --isolation subprocess
+
+    # In a new terminal, start the SuperExec process with the following command:
+    $ flower-superexec \
+        --insecure \
+        --executor flwr.superexec.deployment:executor \
+        --executor-config 'superlink="127.0.0.1:9091"'
```
For 1.13, we have deprecated `flower-superexec`. The two commands users need to use are `flower-superlink` and `flower-supernode`.

dstripelis replied:
Thank you for pointing this out. Removed.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -210,6 +264,30 @@ Deployment
+         --isolation subprocess \
+         --root-certificates <your-ca-cert-filepath>
+
+    # In a new terminal, start the SuperExec process with the following command:
+    $ flower-superexec \
+        --executor flwr.superexec.deployment:executor \
+        --executor-config 'superlink="127.0.0.1:9091"' \
+        --root-certificates <your-ca-cert-filepath>
```
See above comment.

dstripelis replied:
Thank you for pointing this out. Removed.

chongshenng commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -210,6 +264,30 @@ Deployment
+        --executor-config 'superlink="127.0.0.1:9091"' \
+        --root-certificates <your-ca-cert-filepath>
+
+    # Flower up to v1.12
```
We also supported `flwr run` from 1.10 onwards. A good reference is our [releases page](https://github.com/adap/flower/releases).

dstripelis replied:
Seems better to only keep supernode and superlink in the Deploy section. Adjusted accordingly.

jafermarq commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -180,67 +214,95 @@ Deployment
+    $ flower-supernode \
+         --insecure \
+         --superlink 127.0.0.1:9092 \
+         --node-config "partition-id=0 num-partitions=2" \
```
```suggestion
         --node-config "..." \
```

jafermarq replied:
I wouldn't use those `partition-id=0 num-partitions=2` since likely someone using `flower-supernode` doesn't have to think about partitions.

dstripelis replied:
True! Removed partitions and added ellipsis (...).

jafermarq commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -180,67 +214,95 @@ Deployment
+    $ flower-supernode \
+         --insecure \
+         --superlink 127.0.0.1:9092 \
+         --node-config "partition-id=1 num-partitions=2" \
```
```suggestion
         --node-config "..." \
```

jafermarq replied:
same as in my previous comment. Alternatively we could replace it as `--node-config "path/to/local/data"` or something like that?

dstripelis replied:
Seems better to keep the ellipsis, since different users might interpret this differently.

jafermarq reviewed (commented):
Should we also remove the `requirements.txt` comment at the top of the document as well as the "Using Poetry" section? Although people can choose to use them, neither of those will work in a deployment setup that relies on the FAB-delivery mechanism. Because of this I think it's better to remove them.

Another point, I'm wondering if it would be better to split the code blocks for `ClientApp`, `ServerApp`, `Simulation CLI` and `Simulation Notebook` into several (one for current method, other for before 1.1x.0, other for the super legacy method). I think that way would be easier to spot the differences.

jafermarq commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -254,62 +316,120 @@ Simulation in CLI
+    :emphasize-lines: 2,9,13,17,21
+
+    # Flower v1.11+
+    # pyproject.toml
+    [tool.flwr.federations.local - sim - gpu]
+    options.num - supernodes = 10
+    options.backend.client - resources.num - cpus = 2
+    options.backend.client - resources.num - gpus = 0.25
```
```suggestion
```

jafermarq replied:
This should be removed as there is no `pyproject.toml` when doing things in a notebook environment

dstripelis replied:
Thank you for pointing this out. Removed. I thought users need to define the configurations in the `pyproject.toml` before starting the notebook, but they can do so from within the notebook.

jafermarq commented on doc/source/how-to-upgrade-to-flower-next.rst:
```diff
@@ -345,6 +464,9 @@ Simulation in a Notebook
         client_resources=backend_config["client_resources"],
     )
 
+For more advanced information regarding Flower simulation please read the
+|flower_how_to_run_simulations_link|_ guide.
```
I'm thinking we could add this in a more prominent way (maybe with `.. tip:`?) at the beginning of both, CLI and notebook simulation sections?

dstripelis replied:
Nice suggestion. Added a small tip just below the different Flower versioning bullet points.

dstripelis commented:
@jafermarq thank you for the insightful feedback. In the latest commit I addressed all comments except for splitting the code blocks for ClientApp, ServerApp, Simulation CLI and Simulation Notebook into several ones. We should do this in the next release since there are many pieces that need to be moved around in order to support this. In the future, we could even have entirely different doc pages for different versions, e.g., `<>/1.13/how-to-upgrade-to-flower-next.html` , `<>/1.14/how-to-upgrade-to-flower-next.html`.

Overall the latest commit does the following:
- Remove using pip header
- Remove using Poetry header AND section
- Added ellipsis in place of partitions
- Removed pyproject.toml config from simulation
- Added tip suggestion in both CLI and notebook simulations
